components/VM_Arm/src/main.c

```diff
@@ -754,66 +787,66 @@ static int generate_fdt(vm_t *vm, const vm_config_t *vm_config,
     }
     fdtgen_keep_nodes_and_disable(context, plat_keep_device_and_disable, ARRAY_SIZE(plat_keep_device_and_disable));
 
+    int num_paths = 0;
+    char **paths = NULL;
+    if (camkes_dtb_get_node_paths) {
+        paths = camkes_dtb_get_node_paths(&num_paths);
+    }
     fdtgen_keep_nodes(context, (const char **)paths, num_paths);
 
+    /* build a DTB in gen_dtb_buf */
     err = fdtgen_generate(context, fdt_ori);
     fdtgen_free_context(context);
     if (err) {
-        ZF_LOGE("Couldn't generate fdt_ori (%d)\n", err);
+        ZF_LOGE("Couldn't generate base DTB, error %d", err);
         return -1;
     }
-    err = fdt_generate_plat_vcpu_node(vm, gen_fdt);
+
+    /* Now the DTB is in gen_dtb_buf and all manipulation must happen there. */
+
+    err = fdt_generate_plat_vcpu_node(vm, gen_dtb_buf);
     if (err) {
-        ZF_LOGE("Couldn't generate plat_vcpu_node (%d)\n", err);
+        ZF_LOGE("Couldn't generate plat_vcpu_node, error %d", err);
         return -1;
     }
 
     /* generate a memory node */
-    err = fdt_generate_memory_node(gen_fdt, vm_config->ram.base,
+    err = fdt_generate_memory_node(gen_dtb_buf, vm_config->ram.base,
                                    vm_config->ram.size);
     if (err) {
         ZF_LOGE("Couldn't generate memory_node (%d)\n", err);
         return -1;
     }
 
-    /* generate a chosen node */
-    err = fdt_generate_chosen_node(gen_fdt, vm_config->kernel_stdout,
-                                   vm_config->kernel_bootcmdline, NUM_VCPUS);
-    if (err) {
-        ZF_LOGE("Couldn't generate chosen_node (%d)\n", err);
+    return 0;
+}
+
+static int vm_dtb_finalize(vm_t *vm, const vm_config_t *vm_config)
+{
+    assert(vm_config->generate_dtb);
+
+#ifdef CONFIG_VM_PCI_SUPPORT
```

Comment:
Why has `if (config_set(CONFIG_VM_PCI_SUPPORT)) {` been removed?

Reply:
I can use this also. It's a leftover from figuring out the dependencies.